Foundation for eventing controller (#9510)
kyma-project/kyma: components/eventing-controller/README.md

# Eventing controller

## Overview

This component contains controllers for various CustomResourceDefinitions related to eventing in Kyma. The controller comes with 2 containers:
- [controller](https://github.com/kubernetes-sigs/controller-runtime)
- [kube-rbac-proxy](https://github.com/brancz/kube-rbac-proxy)

> ## Prerequisites

Reviewer: Also please mention `kustomize`.

Author: Fixed

Author: Added

    ```shell script
    $ make deploy-local

    $ ## To verify all the manifests after the processing by Kustomize without applying to the cluster use make target deploy-local-dry-run    
    $ make deploy-local-dry-run
    ```

## Usage 

This section explains how to use the eventing-controller.

- The eventing-controller comes with the following command line argument flags

    | Flag    | Default Value | Description                                                                                   |
    | ----------------------- | ------------- |---------------------------------------------------------------------------------------------- |
    | metrics-addr            | :8080          | It's address the metric endpoint binds to.
    | enable-leader-election            | false          | It enables leader-election in controller and ensures there is only one active controller. 
    | lease-duration            | 15s          | It's duration the non-leader candidates will wait to force acquire leadership. Valid time units are ns, us (or µs), ms, s, m, h. 

- To install the CustomResourceDefinitions in a cluster run:

>     ```shell script

Reviewer 2: ```suggestion
    ```console
```

or make it a shell script by removing the '$'

Author: Fixed

    $ make install
    ```

- To uninstall the CustomResourceDefinitions in a cluster run:

    ```shell script

>     $ make install

Reviewer 2: ```suggestion
    $ make uninstall
```

Author: Fixed

# Eventing controller

## Overview

This component contains controllers for various CustomResourceDefinitions related to eventing in Kyma. The controller comes with 2 containers:
- [controller](https://github.com/kubernetes-sigs/controller-runtime)
- [kube-rbac-proxy](https://github.com/brancz/kube-rbac-proxy)

## Prerequisites
- Install [ko](https://github.com/google/ko) which is used to build and deploy the controller during local development
- Install [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) which is the base framework for this controller
- Install [kustomize](https://github.com/kubernetes-sigs/kustomize) which lets customize raw, template-free YAML files during local development

### Installation

- To deploy the controller inside a cluster, make sure you have `ko` installed and configured according to the [usage instructions](https://github.com/google/ko#usage), then run:

    ```sh
    make deploy-local

    ## To verify all the manifests after the processing by Kustomize without applying to the cluster use make target deploy-local-dry-run    
    make deploy-local-dry-run
    ```

## Usage 

> This section explains how to use the eventing-controller.
> 
> - The eventing-controller comes with the following command line argument flags

Reviewer 3: ```suggestion
This section explains how to use the Eventing Controller.

- The Eventing Controller comes with the following command line argument flags:
```

    | Flag    | Default Value | Description                                                                                   |
    | ----------------------- | ------------- |---------------------------------------------------------------------------------------------- |

>     | metrics-addr            | :8080          | It's address the metric endpoint binds to.

Reviewer 3: ```suggestion
    | metrics-addr            | :8080          | Address the metric endpoint binds to.
```

Not sure what "it's address the metric endpoint binds to" means, if it's the Eventing Controller address, we can maybe write "Eventing Controller address the metric endpoint binds to"?

>     | enable-leader-election            | false          | It enables leader-election in controller and ensures there is only one active controller. 
>     | lease-duration            | 15s          | It's duration the non-leader candidates will wait to force acquire leadership. Valid time units are ns, us (or µs), ms, s, m, h. 

Reviewer 3: ```suggestion
    | enable-leader-election            | false          | Enables leader-election in controller and ensures there is only one active controller. 
    | lease-duration            | 15s          | Duration the non-leader candidates will wait to force acquire leadership. Valid time units are ns, us (or µs), ms, s, m, h. 
```

> - To install the CustomResourceDefinitions in a cluster run:
> 
>     ```sh
>     make install
>     ```
> 
> - To uninstall the CustomResourceDefinitions in a cluster run:
> 
>     ```sh
>     make uninstall
>     ```
> 
> - To install the sample CustomResources in a cluster run:
> 
>     ```sh
>     make install-samples
>     ```
> 
> - To uninstall the sample CustomResources in a cluster run:

Reviewer 3: ```suggestion
- To install the CustomResourceDefinitions in a cluster, run:

    ```sh
    make install
    ```

- To uninstall the CustomResourceDefinitions in a cluster, run:

    ```sh
    make uninstall
    ```

- To install the sample CustomResources in a cluster, run:

    ```sh
    make install-samples
    ```

- To uninstall the sample CustomResources in a cluster, run:
```

    ```sh
    make uninstall-samples
    ```

## Development

### Project setup

Before running the component, execute the following command once to pull software dependencies and run tests:

```sh
make test
## To download dependencies only
make resolve-local 
```

### Generate code during local development

> More details on scaffolding code using kubebuilder can be found [here](https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/simplified-scaffolding.md). 

> - Add new apis using [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) CLI followed by generating boiler-plate code by executing the following script:

Reviewer 3: ```suggestion
- Add new APIs using [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) CLI followed by generating boilerplate code by executing the following script:
```

    ```sh
    kubebuilder create api --group batch --version v1 --kind CronJob

    make manifests
    ```

> - Update fields in the `spec` of an existing CustomResourceDefinition by modifying the go file for the type i.e. `api/version/<crd>_types.go`. E.g. `api/v1alpha1/subscriptions_types.go` for Subscriptions CRD. After that, execute the following command to generate boiler-plate code:

Reviewer 3: ```suggestion
- Update fields in the `spec` of an existing CustomResourceDefinition by modifying the Go file for the type i.e. `api/version/<crd>_types.go`. For example, `api/v1alpha1/subscriptions_types.go` for Subscriptions CRD. After that, execute the following command to generate boilerplate code:
```

    ```sh
    make manifests
    ```

> - Add the necessary changes manually in the sample CustomResources after updating fields for an existing CustomResourceDefinition inside the folder `config/samples/`. E.g. for subscriptions update the fields manually in `config/samples/eventing_v1alpha1_subscriptioncomponents/eventing-controller/config/crd/bases/eventing.kyma-project.io_subscriptions.yaml.yaml`

Reviewer 3: ```suggestion
- Add the necessary changes manually in the sample CustomResources after updating fields for an existing CustomResourceDefinition inside the folder `config/samples/`. For example, for subscriptions, update the fields manually in `config/samples/eventing_v1alpha1_subscriptioncomponents/eventing-controller/config/crd/bases/eventing.kyma-project.io_subscriptions.yaml.yaml`
```

# Eventing Controller

## Overview

> This component contains controllers for various CustomResourceDefinitions related to eventing in Kyma. The controller comes with 2 containers:
> - [controller](https://github.com/kubernetes-sigs/controller-runtime)

Reviewer 3: ```suggestion
This component contains controllers for various CustomResourceDefinitions related to eventing in Kyma. The controller comes with this container:
- [controller](https://github.com/kubernetes-sigs/controller-runtime)
```

The document says the "controller comes with 2 containers", but only one is mentioned. Is this intentional? I suggest we either add the 2nd one or re-write this.

## Prerequisites
- Install [ko](https://github.com/google/ko) which is used to build and deploy the controller during local development
- Install [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) which is the base framework for this controller

> - Install [kustomize](https://github.com/kubernetes-sigs/kustomize) which lets you customize raw, template-free YAML files during local development

Reviewer 3: ```suggestion
- Install [kustomize](https://github.com/kubernetes-sigs/kustomize) which lets you customize raw, template-free `yaml` files during local development
```

### Installation

- To deploy the controller inside a cluster, make sure you have `ko` installed and configured according to the [usage instructions](https://github.com/google/ko#usage), then run:

    ```sh
    make deploy-local

>     ## To verify all the manifests after the processing by Kustomize without applying to the cluster use make target deploy-local-dry-run    

Reviewer 3: ```suggestion
    ## To verify all the manifests after the processing by Kustomize without applying to the cluster, use make target deploy-local-dry-run    
```

- Install [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) which is the base framework for this controller
- Install [kustomize](https://github.com/kubernetes-sigs/kustomize) which lets you customize raw, template-free YAML files during local development

### Installation

- To deploy the controller inside a cluster, make sure you have `ko` installed and configured according to the [usage instructions](https://github.com/google/ko#usage), then run:

    ```sh
    make deploy-local

    ## To verify all the manifests after the processing by Kustomize without applying to the cluster use make target deploy-local-dry-run    
    make deploy-local-dry-run
    ```

## Usage 

This section explains how to use the Eventing Controller.

- The Eventing Controller comes with the following command line argument flags:

    | Flag    | Default Value | Description                                                                                   |
    | ----------------------- | ------------- |---------------------------------------------------------------------------------------------- |
    | metrics-addr            | :8080          | The address the metric endpoint binds to.

>     | enable-leader-election            | false          | Enables leader-election in controller and ensures there is only one active controller. 

Reviewer 4: "Leader election" should be deleted

Reviewer 5: Done
